code/xcvm/evm/src/libraries/SDK.sol

```diff
@@ -0,0 +1,870 @@
+        uint64 lowBits;
+        (success, newPos, lowBits) = ProtobufLib.decode_uint64(pos, program);
+        value = uint128(highBits) * 2**64 + uint128(lowBits);
+    }
+
+    function _handleUnit(
+        bytes memory program,
+        uint64 pos,
+        address tokenAddress
+    ) internal view returns (uint256 amount, uint64 newPos) {
+        uint64 size;
+        // read ratio message body
+        (size, pos) = _getMessageLength(program, pos);
+        // reading Unit
+        pos = _checkField(program, 1, ProtobufLib.WireType.LengthDelimited, pos);
+        uint128 unit;
+        (unit, pos) = _handleUint128(program, pos);
+        // reading balance type
+        pos = _checkField(program, 2, ProtobufLib.WireType.LengthDelimited, pos);
+
+        uint256 nominator;
+        uint256 denominator;
+        (nominator, denominator, newPos) = _handleRatio(program, pos);
+        uint256 decimals = IERC20Metadata(tokenAddress).decimals();
+        amount = uint256(unit) * (10**decimals) + (nominator * (10**decimals)) / denominator;
+    }
+
+    function _handleRatio(bytes memory program, uint64 pos)
+        internal pure
+        returns (
+            uint256 nominator,
+            uint256 denominator,
+            uint64 newPos
+        )
+    {
+        uint64 size;
+        // read ratio message body
+        (size, pos) = _getMessageLength(program, pos);
+        pos = _checkField(program, 1, ProtobufLib.WireType.LengthDelimited, pos);
+        (nominator, pos) = _handleUint128(program, pos);
+
+        pos = _checkField(program, 2, ProtobufLib.WireType.LengthDelimited, pos);
+        (denominator, newPos) = _handleUint128(program, pos);
+    }
+
+    function _handleAbsolute(bytes memory program, uint64 pos) internal pure returns (uint256 amount, uint64 newPos) {
+        uint64 size;
+        (size, pos) = _getMessageLength(program, pos);
+
+        pos = _checkField(program, 1, ProtobufLib.WireType.LengthDelimited, pos);
+        (amount, newPos) = _handleUint128(program, pos);
+    }
+
+    function _handleAssetAmount(bytes memory program, uint64 pos, address routerAddress) internal view returns (uint256 amount, uint64 newPos) {
+        uint64 size;
+        bool success;
+        uint64 field;
+        ProtobufLib.WireType _type;
+        // read balance message body
+        (size, pos) = _getMessageLength(program, pos);
+        pos = _checkField(program, 1, ProtobufLib.WireType.LengthDelimited, pos);
+
+        address asset;
+        uint128 assetId;
+        (asset, assetId, pos) = _handleAssetId(program, pos, routerAddress);
+
+        (success, pos, field, _type) = ProtobufLib.decode_key(pos, program);
+        require(field == 2 || field == 3, "decode key failed");
+        require(success, "decode key failed");
+        require(_type == ProtobufLib.WireType.LengthDelimited, "decode type is not embedded messages");
+
+        if (field == 2) {
```

Comment:
since we converted `AssetAmount` to have `Balance` now, I guess we should also add absolute here as well which I guess can use `handleBalance`